File: dasc/kernel.py
import logging
from .schemas import Intent, Decision
from .ledger import BitemporalLedger
from .exceptions import (
    DASCError,
    OCCConflictError, 
    PolicyViolationError, 
    TaintDetectedError, 
    EscalationRequired
)

# Configure logging for production-ready observability
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - DASC - %(levelname)s - %(message)s'
)
logger = logging.getLogger("dasc")

class Kernel:
    def __init__(self, current_state_versions: dict = None):
        self.ledger = BitemporalLedger()
        # Mocking the current state of the world for OCC checks
        self.current_state_versions = current_state_versions or {}

    def evaluate(self, intent: Intent, raise_on_failure: bool = False) -> Decision:
        reasons = []
        error_map = {
            "IFC_TAINT_DETECTED": TaintDetectedError,
            "OCC_CONFLICT": OCCConflictError,
            "OCC_HASH_CONFLICT": OCCConflictError,
            "MISSING_COMPENSATION_PLAN_FOR_HIGH_RISK": PolicyViolationError,
            "TIER_4_MANDATORY_ESCALATION": EscalationRequired
        }

        logger.info(f"Evaluating Intent {intent.intent_id} from agent {intent.actor_agent}")

        # 1. IFC / Taint Check
        for ev in intent.evidence:
            if ev.source_type == "untrusted" and intent.risk_tier >= 2:
                msg = f"IFC_TAINT_DETECTED: Untrusted evidence for sensitive action (Tier {intent.risk_tier})"
                reasons.append(msg)
                logger.warning(f"[{intent.intent_id}] {msg}")

        # 2. OCC / TOCTOU Check
        if intent.state_version_vector:
            from .utils import calculate_file_hash
            for artifact, version in intent.state_version_vector.items():
                current_v = self.current_state_versions.get(artifact)
                
                # If version is prefixed with 'hash:', calculate the actual file hash
                if version.startswith("hash:"):
                    expected_hash = version.split("hash:")[1]
                    actual_hash = calculate_file_hash(artifact)
                    if actual_hash and actual_hash != expected_hash:
                        msg = f"OCC_HASH_CONFLICT: {artifact} content changed. Expected {expected_hash[:8]}, got {actual_hash[:8]}"
                        reasons.append(msg)
                        logger.warning(f"[{intent.intent_id}] {msg}")
                elif current_v and current_v != version:
                    msg = f"OCC_CONFLICT: {artifact} changed from {version} to {current_v}"
                    reasons.append(msg)
                    logger.warning(f"[{intent.intent_id}] {msg}")

        # 3. Risk Tier Policy Check
        if intent.risk_tier >= 3 and not intent.compensation_plan:
            msg = "MISSING_COMPENSATION_PLAN_FOR_HIGH_RISK"
            reasons.append(msg)
            logger.warning(f"[{intent.intent_id}] {msg}")

        # 4. Formulate Decision
        if reasons:
            decision = Decision(intent_id=intent.intent_id, status="REJECT", reason_codes=reasons)
            logger.error(f"[{intent.intent_id}] REJECTED. Reasons: {reasons}")
            if raise_on_failure:
                # Raise the first encountered error type
                error_cls = next((error_map[r.split(":")[0]] for r in reasons if r.split(":")[0] in error_map), DASCError)
                raise error_cls(str(reasons), intent_id=intent.intent_id)
        elif intent.risk_tier == 4:
            decision = Decision(intent_id=intent.intent_id, status="ESCALATE", reason_codes=["TIER_4_MANDATORY_ESCALATION"])
            logger.info(f"[{intent.intent_id}] ESCALATED for human approval.")
            if raise_on_failure:
                raise EscalationRequired("Human gate required for Tier 4 action", intent_id=intent.intent_id)
        else:
            decision = Decision(intent_id=intent.intent_id, status="COMMIT")
            logger.info(f"[{intent.intent_id}] COMMITTED successfully.")

        # 5. Record and Return
        self.ledger.log_decision(intent, decision)
        return decision

File: dasc-node/src/kernel.ts
import { Intent, Decision, DASCStatus } from "./types.js";
import { BitemporalLedger } from "./ledger.js";
import { NotificationManager } from "./notifications.js";

export type PolicyFunc = (intent: Intent) => { pass: boolean; reason?: string };

export class Kernel {
  private ledger: BitemporalLedger;
  private notifier: NotificationManager;
  private current_state_versions: Record<string, string>;
  private policies: PolicyFunc[] = [];

  constructor(current_state_versions: Record<string, string> = {}, dbPath?: string) {
    this.ledger = new BitemporalLedger(dbPath);
    this.notifier = new NotificationManager();
    this.current_state_versions = current_state_versions;
  }

  registerPolicy(policy: PolicyFunc) {
    this.policies.push(policy);
  }

  async approveIntent(intent_id: string, approved: boolean, approver: string): Promise<Decision> {
    const history = this.ledger.getHistory();
    const target = (history as any[]).find((h: any) => h.intent_id === intent_id);

    if (!target) throw new Error("Intent not found");
    if (target.status !== "ESCALATE") throw new Error("Only escalated intents can be approved");

    const status: DASCStatus = approved ? "COMMIT" : "REJECT";
    const decision: Decision = {
      intent_id,
      status,
      reason_codes: [`MANUAL_${status}_BY_${approver}`],
      suggestions: [],
      timestamp: new Date().toISOString()
    };

    this.ledger.logDecision(JSON.parse(target.intent_json), decision);
    return decision;
  }

  evaluate(intent: Intent): Decision {
    const reasons: string[] = [];
    
    // 1. IFC / Taint Check
    for (const ev of intent.evidence) {
      if (ev.source_type === "untrusted" && intent.risk_tier >= 2) {
        reasons.push(`IFC_TAINT_DETECTED: Untrusted evidence for Tier ${intent.risk_tier}`);
      }
    }

    // 2. OCC / Version Check
    if (intent.state_version_vector) {
      for (const [artifact, version] of Object.entries(intent.state_version_vector)) {
        const current_v = this.current_state_versions[artifact];
        if (current_v && current_v !== version) {
          reasons.push(`OCC_CONFLICT: ${artifact} changed from ${version} to ${current_v}`);
        }
      }
    }

    // 3. Risk Tier Policy
    if (intent.risk_tier >= 3 && !intent.compensation_plan) {
      reasons.push("MISSING_COMPENSATION_PLAN_FOR_HIGH_RISK");
    }

    // 4. Custom Policies
    for (const policy of this.policies) {
      const result = policy(intent);
      if (!result.pass) {
        reasons.push(`CUSTOM_POLICY_VIOLATION: ${result.reason}`);
      }
    }

    // 5. Formulate Decision
    let status: DASCStatus = "COMMIT";
    if (reasons.length > 0) {
      status = reasons.some(r => r.includes("TIER_4") || r.includes("HIGH_RISK")) ? "ESCALATE" : "REJECT";
    } else if (intent.risk_tier === 4) {
      status = "ESCALATE";
      reasons.push("TIER_4_MANDATORY_ESCALATION");
    }

    const suggestions: string[] = [];
    if (status !== "COMMIT") {
      if (reasons.includes("MISSING_COMPENSATION_PLAN_FOR_HIGH_RISK")) {
        suggestions.push("Provide a 'compensation_plan' (e.g. {undo: 'revert'})");
      }
      if (status === "ESCALATE") {
        suggestions.push("Human approval required via Control Plane.");
      }
    }

    const decision: Decision = {
      intent_id: intent.intent_id,
      status,
      reason_codes: reasons,
      suggestions,
      timestamp: new Date().toISOString()
    };

    // 6. Record and potentially notify
    this.ledger.logDecision(intent, decision);
    
    if (status !== "COMMIT") {
        this.notifier.notify(intent, decision).catch(console.error);
    }

    return decision;
  }
}
